plugin/semantic.py

```diff
@@ -0,0 +1,232 @@
+[MODIFICATION_SCOPES.update({k:v + ', semantic.'+ k + '.modification.lsp'})   for k,v in MODIFICATION_SCOPES.items()]
+[DEFAULT_LIB_SCOPES.update( {k:v + ', semantic.'+ k + '.defaultLibrary.lsp'}) for k,v in DEFAULT_LIB_SCOPES.items()]
+
+SEMANTIC_SCOPES = {
+            ''              :NON_MODIFIER_SCOPES,# if no modifiers are provided
+
+            "declaration"   :DECLARATION_SCOPES,
+            "definition"    :DECLARATION_SCOPES,
+
+            "readonly"      :{'variable': 'constant.other.lsp, variable.other.constant.lsp, semantic.variable.readonly.lsp'},
+            "static"        :STAIC_SCOPES, # these are temporary, should be filled with real scopes
+            "deprecated"    :DEPRECATED_SCOPES,
+            "abstract"      :ABSTRACT_SCOPES,
+            "async"         :ASYNC_SCOPES,
+            "modification"  :MODIFICATION_SCOPES,
+            "documentation" :{'comment':'comment.block.documentation'},
+            "defaultLibrary":DEFAULT_LIB_SCOPES
+        }
+
+
+def get_semantic_scope_from_modifier(encoded_token:list, semantic_tokens_legends:dict) -> str:
+    token_int    = encoded_token[3]
+    modifier_int = encoded_token[4]
+
+    token_type = semantic_tokens_legends['tokenTypes'][token_int]
+
+    if not modifier_int:
+        try: return(SEMANTIC_SCOPES[''][token_type])
+        except KeyError: pass #print('scope not defined for tokenType: ', token_type)
+
+    modifier_binary = [int(x) for x in reversed(bin(modifier_int)[2:])]
+    modifiers = [semantic_tokens_legends['tokenModifiers'][i] for i in range(len(modifier_binary)) if i]
+
+    # print('tokenType: ' + token_type + '-- modifiers: ',modifiers)
+    scopes = []
+    for modifier in modifiers:
+        scope = None
+        try: scope = SEMANTIC_SCOPES[modifier][token_type]
+        except KeyError: pass #print('scope not defined for modifier/tokenType: ', modifier+'/'+token_type)
+
+        if scope and scope not in scopes:
+            scopes.append(scope)
+
+    if not scopes:
+        try: return(SEMANTIC_SCOPES[''][token_type])
+        except KeyError: pass #print('scope not defined for tokenType: ', token_type)
+    else:
+        return ' '.join(scopes)
+
+
+class LspSemanticCommand(LspTextCommand):
+
+    capability = 'semanticTokensProvider'
+
+    def __init__(self, view: sublime.View) -> None:
+        super().__init__(view)
+        self._base_dir = None   # type: Optional[str]
+
+    def run(
+        self,
+        edit: sublime.Edit,
+        only_diagnostics: bool = False
+    ) -> None:
+        if 'background' not in self.view.style_for_scope('meta.semantic-token.lsp'):
+            print('''
+        Color Scheme :
+        Error: Scope not defined in ColorSchme
+        
+        Add the following to your color scheme
+        
+        "rules":
+        [
+            {
+                "scope": "meta.semantic-token.lsp",
+                "background": "#00000001"
```

Comment:
I think `#01010101` will cause less problem. This doesn't work for me as my color scheme background is perfect black (i.e., `#000000`).

Reply:
I don't even know what it does, I'm guessing it's setting the lower bit,
jwortmann suggested using it, so I copy/pasta it.

Reply:
To make `add_region()` able to change to foreground color, the background color for `add_region()` has to be different with the background color in the user's color scheme. To make the difference subtle so human eyes can't even distinguish it, people tend to just change the background color a little bit. But changing the alpha of the color doesn't work, at least on my side.

Reply:
Yes, you are absolutely right @jfcherng, if the background is already completely black, then `#00000001` won't work. This was more of an example, and we should better describe it in the docs, similar to the unnecessary/deprecated diagnostic tags https://lsp.sublimetext.io/customization/#diagnostics.

Reply:
> To make `add_region()` able to change the foreground color, the background color for `add_region()` has to be different with the `--background` color in the user's color scheme ([ST core bug](https://github.com/sublimehq/sublime_text/issues/817)?). To make the difference subtle so human eyes can't even distinguish it, plugin devs tend to just change the `--background` color a little bit. But changing the alpha of the color doesn't work, at least on my side.
> 
> That said, if you use
> 
> ```js
> {
>     "scope": "meta.semantic-token.lsp",
>     "background": "#00000001",
> }
> ```
> 
> then your trick won't work for people who use `#000000` as their `--background` color.
> 
> Similarly, if you use
> 
> ```js
> {
>     "scope": "meta.semantic-token.lsp",
>     "background": "#01010101",
> }
> ```
> 
> then your trick won't work for people who use `#010101` as their `--background` color. But there should be less people use `#010101` as their `--background` color.

I knew this much but, I wasn't entirely sure adding "background": "#00000001",  is masking the existing  background color or what?

Reply:
Anyway, it should be a color which is close to user's `--background` but not exactly the same, for best UX.

Reply:
by masking I meant bit-masking

Reply:
> by masking I meant bit-masking

no. that trailing 01 is alpha.